nuget/lib/dependabot/nuget/nuget_config_credential_helpers.rb

```diff
@@ -0,0 +1,69 @@
+# typed: true
+# frozen_string_literal: true
+
+module Dependabot
+  module Nuget
+    module NuGetConfigCredentialHelpers
+      def self.user_nuget_config_path
+        home_directory = Dir.home
+        File.join(home_directory, ".nuget", "NuGet", "NuGet.Config")
+      end
+
+      def self.temporary_nuget_config_path
+        user_nuget_config_path + "_ORIGINAL"
+      end
+
+      def self.add_credentials_to_nuget_config(credentials)
+        return unless File.exist?(user_nuget_config_path)
+
+        nuget_credentials = credentials.select { |cred| cred["type"] == "nuget_feed" }
+        return if nuget_credentials.empty?
+
+        File.rename(user_nuget_config_path, temporary_nuget_config_path)
+
+        package_sources = []
+        package_source_credentials = []
+        nuget_credentials.each_with_index do |c, i|
+          source_name = "credentialed_source_#{i + 1}"
+          package_sources << "    <add key=\"#{source_name}\" value=\"#{c.fetch('url')}\" />"
+          package_source_credentials << "    <#{source_name}>"
+          package_source_credentials << "      <add key=\"Username\" value=\"user\" />"
+          package_source_credentials << "      <add key=\"ClearTextPassword\" value=\"#{c.fetch('token')}\" />"
+          package_source_credentials << "    </#{source_name}>"
+        end
+
+        nuget_config = <<~NUGET_XML
+          <?xml version="1.0" encoding="utf-8"?>
+          <configuration>
+            <packageSources>
+          #{package_sources.join("\n")}
+            </packageSources>
+            <packageSourceCredentials>
+          #{package_source_credentials.join("\n")}
+            </packageSourceCredentials>
+          </configuration>
+        NUGET_XML
+        File.write(user_nuget_config_path, nuget_config)
+      end
+
+      def self.restore_user_nuget_config
+        return unless File.exist?(temporary_nuget_config_path)
+
+        File.delete(user_nuget_config_path)
+        File.rename(temporary_nuget_config_path, user_nuget_config_path)
+      end
+
+      # rubocop:disable Lint/SuppressedException
+      def self.patch_nuget_config_for_action(credentials, action)
+        add_credentials_to_nuget_config(credentials)
+        begin
+          action.call
+        rescue StandardError
+        ensure
+          restore_user_nuget_config
+        end
+      end
+      # rubocop:enable Lint/SuppressedException
```

Comment:
It might be a bit more idiomatic/ruby-ish to have this method accept a block ([this example](https://github.com/dependabot/dependabot-core/blob/763f4449c2898c6ed3bd6c9c36beabde6f33f32b/common/lib/dependabot/shared_helpers.rb#L254) comes to mind) by tossing an `&` in front of the second parameter

```ruby
def self.with_patched_nuget_config(credentials, &action)
  add_credentials_to_nuget_config(credentials)
  begin
    action.call
  rescue StandardError
  ensure
    restore_user_nuget_config
  end
end
```

which would change the calling code [above](https://github.com/dependabot/dependabot-core/pull/8748/files#diff-ed307fbccf4b647ac64a7cde4b3449b7a0fd61fe90247280449ea2fbcb29404dR89) to this:

```ruby
NuGetConfigCredentialHelpers.with_patched_nuget_config(credentials) do
  output = SharedHelpers.run_shell_command(command, fingerprint: fingerprint)
  puts output
end
```

Reply:
Thank you, this is great!  I didn't know Ruby had this feature.  I'll update this shortly.

Reply:
Done.